File: server/validations/val.go
package val

import (
	"fmt"
	"net/mail"
	"regexp"
)

var (
	isValidUsername = regexp.MustCompile(`^[a-zA-Z\s]+$`).MatchString
)

func ValidateString(value string, min int, max int) error {
	n := len(value)
	if n < min || n > max {
		return fmt.Errorf("string should have length between %d and %d", min, max)
	}
	return nil
}

func ValidateUserName(value string) error {
	if err := ValidateString(value, 3, 50); err != nil {
		return err
	}

	if ok := isValidUsername(value); !ok {
		return fmt.Errorf("user name should contain only letters, capital letters and spaces")
	}

	return nil
}

func ValidatePassword(value string) error {
	return ValidateString(value, 6, 100)
}

func ValidateEmail(value string) error {
	if err := ValidateString(value, 3, 200); err != nil {
		return err
	}

	if _, err := mail.ParseAddress(value); err != nil {
		return err
	}

	return nil
}

File: client/validations/val.ts
const emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
const usernameRegex = /`^[a-zA-Z\s]+$`/

export function validateString(value: string, min: number, max: number): string {
    if (value.length < min || value.length > max) return `should have length between ${min} and ${max}`
    return ""
}

export function validateEmail(value: string): string {
    if (!emailRegex.test(value)) return "not a valid email"
    return validateString(value, 3, 200)
}

export function validatePassword(value: string): string {
    return validateString(value, 6, 100)
}

export function validateUsername(value: string): string {
    if (!usernameRegex.test(value)) return "user name should contain only letters, capital letters and spaces"
    return validateString(value, 3, 50)
}